### PR TITLE
disable gtFoldExprConst for conversion from negative double to unsigned long

### PR DESCRIPTION
**Pull request by hseok-oh**
Related issue: #6667 

In ARM architecture, negative double is converted to zero when target is unsigned integer.
Reference comment and implementation:  `FloatingPointUtils::convertDoubleToUInt64(double d)` in jit/utils.cpp

### CODE REVIEW

**Comment by jkotas** [comment]
@dotnet/jit-contrib Would it be better to unconditionally disable gtFoldExprConst in the JIT where it gives undefined or imprecise results instead? We are going to be chasing mismatches between JIT and JIT helper implementations forever otherwise.

**Comment by RussKeldorph** [comment]
@jkotas' Question valid question aside, I don't think I would change the behavior of a method called "FastDbl2Lng" so fundamentally without at least changing its name.  Alternatively, I haven't gone through all the cases to be certain, but it feels like you ought to be modifying the JIT_Dbl2U* helpers directly and adding #ifdefs to detect the quirkiness of MSVC (#ifdef _MSC_VER?) versus other compilers, which may already not require the range-checking gymnastics currently implemented in the helpers.

**Comment by hseok-oh** [comment]
I fundamentally agree with your opinion modifying JIT_Dbl2U* helpers directly. But I modified FastDbl2Lng because there was no helper functions in `jithelpers.cpp` that uses "#ifdefs" in helper functions for different architectures.
If there is no problem, I'll modify helpers directly.

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by hseok-oh** [comment]
@dotnet-bot test OSX x64 Checked Build and Test please

**Comment by hseok-oh** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

**Comment by hseok-oh** [comment]
Modify helper function `JIT_Dbl2ULng` directly. 
Other helpers such as `JIT_Dbl2UIntOvf` and `JIT_Dbl2ULngOvf` did not modified, because these function throw exception if the result of conversion is negative.

@jkotas , @RussKeldorph could you please have a look?

File: src/vm/jithelpers.cpp [hunk 1]
```diff
@@ -613,13 +613,36 @@ HCIMPL1_V(UINT64, JIT_Dbl2ULng, double val)
 
     const double two63  = 2147483648.0 * 4294967296.0;
     UINT64 ret;
+#ifdef _TARGET_XARCH_
```
Comment (jkotas) [inline src/vm/jithelpers.cpp]: This implementation works fine on Windows everywhere. We should keep using instead of the slower one below.

**Comment by jkotas** [comment]
I am still wondering that a better fix would be to remove this `#ifdef _TARGET_XARCH_` ifdef in the JIT: https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.cpp#L11985, and leave the JIT helper alone.

File: src/vm/jithelpers.cpp (hunk 1 above)
Comment (hseok-oh) [inline src/vm/jithelpers.cpp]: @jkotas Could you recommend proper flag? We need to keep using upper one on Windows everywhere, and Linux for x86/x64.

**Comment by RussKeldorph** [comment]
(@jkotas I was very confused for a bit since the line you referenced has moved recently.  I edited your comment with the stable link, which you can get by pressing 'y' while viewing the unstable link in the browser.)

@hseok-oh FWIW, I agree with @jkotas about gtFoldExprConst.  I'm skeptical there is a code quality advantage to folding these cases which are problematic to emulate because the behavior is unspecified in both C++ and IL.

**Comment by hseok-oh** [comment]
@jkotas, @RussKeldorph I also agree with you about `gtFoldExprConst`. The most important thing is that we should offer same result always by casting from negative floating point to unsigned integer in CoreCLR. 
I think these casting will be not used frequently and optimization gives us limited benefit, compared to general casting.

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by hseok-oh** [comment]
@jkotas , @RussKeldorph  I have modified `gtFoldExprConst` to disable when convert to unsigned integer from negative double, instead of modifying helper function.

File: src/jit/gentree.cpp [hunk 2]
```diff
@@ -12042,49 +12042,17 @@ GenTreePtr Compiler::gtFoldExprConst(GenTreePtr tree)
...
-                                    if (op1->gtType == TYP_FLOAT)
-                                        d1 = forceCastToFloat(d1); // it's only !_finite() after this conversion
-                                    goto CNS_DOUBLE;
-                                default:
-                                    unreached();
-                            }
-#endif //!_TARGET_XARCH_
                         }
 
+                        if (d1 < 0.0) {
```
Comment (RussKeldorph) [inline src/jit/gentree.cpp]: Should this be if (d1 <= 1.0)?  I believe this is consistent with other places that test the same thing.

Also please place opening brace on next line.  (Source formatting tool is close...)

File: src/jit/gentree.cpp [hunk 3]
```diff
@@ -12042,49 +12042,17 @@ GenTreePtr Compiler::gtFoldExprConst(GenTreePtr tree)
...
+                            if (tree->CastToType() == TYP_CHAR || tree->CastToType() == TYP_UBYTE ||
```
Comment (RussKeldorph) [inline src/jit/gentree.cpp]: Is varTypeIsUnsigned appropriate here?

@dotnet/jit-contrib Please review and comment.  It seems safe (perf-wise) and appropriate to me to disallow folding on all architectures when the result is un-/implementation-defined, but I'm curious if others have concerns.

Comment (JosephTremoulet) [inline src/jit/gentree.cpp]: >  It seems safe (perf-wise) and appropriate to me to disallow folding on all architectures when the result is un-/implementation-defined

I'm inclined to agree.

File: src/jit/gentree.cpp (hunk 2 above)
Comment (hseok-oh) [inline src/jit/gentree.cpp]: @RussKeldorph Is it typo `if (d1 <= -1.0)` ?

**Comment by hseok-oh** [comment]
@jkotas , @RussKeldorph I've modified gtFoldExprConst by using `varTypeIsUnsigned`.

**Comment by hseok-oh** [comment]
@jkotas , @RussKeldorph PTAL

**Comment by jkotas** [comment]
LGTM

**Comment by hseok-oh** [comment]
@jkotas @RussKeldorph @JosephTremoulet ready to get merged.
Does it have another issues to get merged?

**Comment by JosephTremoulet** [comment]
LGTM, but probably worth adding a comment indicating that we don't fold these cases because the result is unspecified per ECMA and the native math doing the fold doesn't match the run-time computation on all platforms and we want the behavior to be the same with or without folding.

**Comment by hseok-oh** [comment]
@dotnet/jit-contrib PTAL

**Comment by JosephTremoulet** [comment]
LGTM

**Pull request merged**